src/codec.rs

```diff
@@ -781,11 +784,37 @@ impl Decode for Vec<u8> {
 	fn decode<I: Input>(input: &mut I) -> Option<Self> {
 		<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
 			let len = len as usize;
-			let mut vec = vec![0; len];
-			if input.read(&mut vec[..len]) != len {
-				None
+
+			if len < MAXIMUM_PREALLOCATION_SIZE {
+				let mut vec = vec![0; len]; // len is ok here
+				if input.read(&mut vec[..len]) != len {
+					None
+				} else {
+					Some(vec)
+				}
 			} else {
-				Some(vec)
+				// if len is considered too much for preallocation then use dynamic allocation
+				let mut vec = Vec::new();
+				let buffer_len = MAXIMUM_PREALLOCATION_SIZE;
+				let mut buffer = vec![0; buffer_len];
+				let mut total_len_read = 0;
+
+				loop {
+					let len_read = input.read(&mut buffer[..buffer_len]);
```

Comment:
Okay, this algorithm is not correct. You need to do `len -= len_read` in each round. Always read `buffer[..min(buffer_len, len)]`.
I think you understand the rest.

Reply:
oh yes, I still forget about this. done